chainer/_link.py

```diff
@@ -0,0 +1,373 @@
+    """
+
+    _device = _cpu.CpuDevice()
+
+    # Indicates the object is mixed-device.
+    _MIXED_DEVICE = object()
+
+    def __init__(self):
+        # Check and store overridden to_device family method names.
+        overridden_to_methods = tuple([
+            m for m in ('to_device', 'to_cpu', 'to_gpu', 'to_intel64')
+            if _is_to_device_method_overridden(self, m)])
+        if overridden_to_methods:
+            # to_device() cannot be overridden.
+            if 'to_device' in overridden_to_methods:
+                raise TypeError(
+                    'to_device cannot be overridden (class {}).'.format(
+                        self.__class__))
+            # Overriding to_cpu/to_gpu/to_intel64 causes a warning.
+            warnings.warn(
+                'Overriding method(s) [{}] of class {} is deprecated. '
+                'Override `visit_device_residents` instead.'.format(
+                    ', '.join(overridden_to_methods),
+                    self.__class__),
+                DeprecationWarning)
+        self._overridden_to_methods = overridden_to_methods
+
+    def visit_device_residents(self, visitor):
+        """Applies the visitor to all the device objects in this instance."""
+        raise NotImplementedError(
+            'Concrete implementation must override this method to describe '
+            'the logic to traverse device objects.')
+
+    @property
+    def device(self):
+        """Returns the sole device"""
+        if self._device is None:
+            visitor = _GetSoleDeviceVisitor()
+            self.visit_device_residents(visitor)
+            sole_device = visitor.sole_device
+            if sole_device is None:
+                self._device = self._MIXED_DEVICE
+            else:
+                self._device = sole_device
+
+        if self._device is self._MIXED_DEVICE:
+            return None
+        return self._device
+
+    @property
+    def xp(self):
+        # type: () -> types.Xp
+        """Array module for this link.
+
+        Depending on which of CPU/GPU this link is on, this property returns
+        :mod:`numpy` or :mod:`cupy`.
+
+        """
+        device = self.device
+        if device is None:
+            return None
+        return device.xp
+
+    def to_cpu(self):
+        # type: () -> 'DeviceResidentMixin'
+        """Copies parameter variables and persistent values to CPU.
+
+        This method does not handle non-registered attributes. If some of such
+        attributes must be copied to CPU, the link implementation must
+        override :meth:`Link.to_device` to do so.
+
+        Returns: self
+
+        """
+        visitor = _ToDeviceVisitor(
+            backend.CpuDevice(),
+            entry_method_info=('to_cpu', {}),
+            skip_between_cupy_devices=True)
+        self.__to_device(visitor)
+        return self
+
+    def to_gpu(
+            self,
+            device=None  # type: tp.Optional[types.CudaDeviceSpec]
```

Comment:
It is a matter of taste, but I prefer putting a trailing comma here (to keep the diff simple on adding a new argument).

Reply:
I agree. Fixed.